crates/primitives/src/l1/payload.rs

```diff
@@ -0,0 +1,174 @@
+    commitment: Buf32,
+}
+
+impl BlobSpec {
+    /// The target we expect the DA payload to be stored on.
+    pub fn dest(&self) -> BlobDest {
+        self.dest
+    }
+
+    /// Commitment to the payload.
+    pub fn commitment(&self) -> &Buf32 {
+        &self.commitment
+    }
+
+    fn new(dest: BlobDest, commitment: Buf32) -> Self {
+        Self { dest, commitment }
+    }
+}
+
+/// Data that is submitted to L1. This can be DA, Checkpoint, etc.
+#[derive(Clone, Debug, Eq, PartialEq, Arbitrary, BorshDeserialize, BorshSerialize)]
+pub struct L1Payload {
+    data: Vec<u8>,
+    payload_type: L1PayloadType,
+}
+
+impl L1Payload {
+    pub fn new(data: Vec<u8>, payload_type: L1PayloadType) -> Self {
+        Self { data, payload_type }
+    }
+
+    pub fn new_checkpoint(data: Vec<u8>) -> Self {
+        Self::new(data, L1PayloadType::Checkpoint)
+    }
+
+    pub fn new_da(data: Vec<u8>) -> Self {
+        Self::new(data, L1PayloadType::Da)
+    }
+
+    pub fn data(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn payload_type(&self) -> &L1PayloadType {
+        &self.payload_type
+    }
+
+    pub fn hash(&self) -> Buf32 {
+        let mut buf = self.data.clone();
+        buf.extend(borsh::to_vec(&self.payload_type).expect("Could not serialize payload type"));
+        hash::raw(&buf)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Arbitrary, BorshDeserialize, BorshSerialize)]
+pub enum L1PayloadType {
+    Checkpoint,
+    Da,
+}
+
+/// Intent produced by the EE on a "full" verification, but if we're just
+/// verifying a proof we may not have access to this but still want to reason
+/// about it.
+///
+/// These are never stored on-chain.
+#[derive(Clone, Debug, Eq, PartialEq, Arbitrary, BorshDeserialize, BorshSerialize)]
+pub struct PayloadIntent {
+    /// The destination for this payload.
+    dest: BlobDest,
+
+    /// Commitment to the payload.
+    commitment: Buf32,
+
+    /// Blob payload.
+    payload: L1Payload,
+}
+
+impl PayloadIntent {
+    pub fn new(dest: BlobDest, commitment: Buf32, payload: L1Payload) -> Self {
+        Self {
+            dest,
+            commitment,
+            payload,
+        }
+    }
+
+    /// The target we expect the DA payload to be stored on.
+    pub fn dest(&self) -> BlobDest {
+        self.dest
+    }
+
+    /// Commitment to the payload, which might be context-specific. This
+    /// is conceptually unrelated to the payload ID that we use for tracking which
+    /// payloads we've written in the L1 writer bookkeeping.
+    pub fn commitment(&self) -> &Buf32 {
+        &self.commitment
+    }
+
+    /// The payload that matches the commitment.
+    pub fn payload(&self) -> &L1Payload {
+        &self.payload
+    }
+
+    /// Generates the spec from the relevant parts of the payload intent that
+    /// uniquely refers to the payload data.
+    pub fn to_spec(&self) -> BlobSpec {
+        BlobSpec::new(self.dest, self.commitment)
+    }
```

Comment:
Should be a new type `PayloadSpec`.

Reply:
I've added a new type `PayloadSpec` and kept `BlobSpec` as is. As of now both of these have the same structure.
Also, `BlobSpec` has `dest` which is `PayloadDest`, I am not sure if it needs to be another `BlobDest`, which I think is redundant. If you have opinions on this and needs change, please be exact.